diff -ur org-20150907.orig/org.el org-20150907/org.el
--- org-20150907.orig/org.el	2015-09-10 10:58:33.282464843 +0100
+++ org-20150907/org.el	2015-09-10 11:00:20.807925505 +0100
@@ -5921,7 +5921,14 @@
 	      (cond
 	       ((and lang (not (string= lang "")) org-src-fontify-natively)
 		(org-src-font-lock-fontify-block lang block-start block-end)
-		(add-text-properties beg1 block-end '(src-block t)))
+                (mapc (lambda (ov)
+                        (if (eq (overlay-get ov 'face) 'org-block-background)
+                            (delete-overlay ov)))
+                      (overlays-at (/ (+ beg1 block-end) 2)))
+                ;; add a background overlay
+                (setq ovl (make-overlay beg1 block-end))
+                (overlay-put ovl 'face 'org-block-background)
+                (overlay-put ovl 'evaporate t)) ; make it go away when empty
 	       (quoting
 		(add-text-properties beg1 (min (point-max) (1+ end1))
 				     '(face org-block))) ; end of source block
@@ -22367,8 +22374,10 @@
 When INSIDE is non-nil, don't consider we are within a src block
 when point is at #+BEGIN_SRC or #+END_SRC."
   (let ((case-fold-search t) ov)
-    (or (and (eq (get-char-property (point) 'src-block) t))
-	(and (not inside)
+    (or (and (setq ov (overlays-at (point)))
+             (memq 'org-block-background
+                   (overlay-properties (car ov))))
+        (and (not inside)
 	     (save-match-data
 	       (save-excursion
 		 (beginning-of-line)
diff -ur org-20150907.orig/org-faces.el org-20150907/org-faces.el
--- org-20150907.orig/org-faces.el	2015-09-10 10:58:33.278465091 +0100
+++ org-20150907/org-faces.el	2015-09-10 11:00:41.574684521 +0100
@@ -479,6 +479,9 @@
   :group 'org-faces
   :version "22.1")
 
+(defface org-block-background '((t ()))
+  "Face used for the source block background.")
+
 (defface org-meta-line
   (org-compatible-face 'font-lock-comment-face nil)
   "Face for meta lines starting with \"#+\"."
